File: TS/系统/02．物品系统/15．装备技能/00．物品/25．咆哮之心.ts
/** @noSelfInFile */


const { debugLogForce } = require("lib.扩展函数.自定义扩展函数.03．调试输出") as {
  debugLogForce: (this: void, module: string, ...args: any[]) => void;
};

const jass = require("jass.common") as any;

const { createTimedEffect } = require("lib.扩展函数.封装函数.01．通用工具.03．特效") as {
  createTimedEffect: (this: void, modelPath: string, x: number, y: number, z?: number, duration?: number) => any;
};
const { SGSS_SetState } = require("lib.扩展函数.Star扩展函数.00．SGSS") as {
  SGSS_SetState: (this: void, unit: any, id: number, value: number) => void;
};

const GetItemTypeId = jass.GetItemTypeId as (item: any) => number;
const GetUnitX = jass.GetUnitX as (unit: any) => number;
const GetUnitY = jass.GetUnitY as (unit: any) => number;
const GetUnitState = jass.GetUnitState as (unit: any, state: any) => number;
const ConvertUnitState = jass.ConvertUnitState as (stateId: number) => any;
const R2I = jass.R2I as (value: number) => number;
const UnitDamageTarget = jass.UnitDamageTarget as (source: any, target: any, amount: number, attack: boolean, ranged: boolean, attackType: any, damageType: any, weaponType: any) => boolean;
const CreateTimer = jass.CreateTimer as () => any;
const TimerStart = jass.TimerStart as (timer: any, timeout: number, periodic: boolean, callback: (this: void) => void) => void;
const GetExpiredTimer = jass.GetExpiredTimer as () => any;
const GetHandleId = jass.GetHandleId as (handle: any) => number;
const DestroyTimer = jass.DestroyTimer as (timer: any) => void;
const ATTACK_TYPE_NORMAL = jass.ATTACK_TYPE_NORMAL as any;
const DAMAGE_TYPE_MIND = jass.DAMAGE_TYPE_MIND as any;
const WEAPON_TYPE_WHOKNOWS = jass.WEAPON_TYPE_WHOKNOWS as any;

import type { 物品技能事件上下文 } from "../03．主动技能/03．物品使用触发/01．物品使用触发常量";
import { 咆哮之心物品ID } from "../03．主动技能/00．公共/01．主动技能物品ID";
import { 咆哮之心配置 } from "../03．主动技能/03．物品使用触发/00．物品使用触发配置";

interface 咆哮之心上下文 {
  施法单位: any;
  目标单位: any;
  附加攻击: number;
  次数: number;
}

const 咆哮之心表: Record<number, 咆哮之心上下文 | undefined> = {};

function 是否为咆哮之心(this: void, 物品: any): boolean {
  if (物品 == null || 物品 === 0) return false;
  return GetItemTypeId(物品) === 咆哮之心物品ID;
}

function on咆哮之心周期(this: void): void {
  const timer = GetExpiredTimer();
  const timerID = GetHandleId(timer);
  const 上下文 = 咆哮之心表[timerID];
  if (上下文 == null) {
    DestroyTimer(timer);
    return;
  }
  if (上下文.次数 >= 咆哮之心配置.次数) {
    SGSS_SetState(上下文.目标单位, 1, -上下文.附加攻击);
    delete 咆哮之心表[timerID];
    DestroyTimer(timer);
    return;
  }
  上下文.次数 += 1;
  createTimedEffect(咆哮之心配置.特效路径, GetUnitX(上下文.目标单位), GetUnitY(上下文.目标单位), 0, 咆哮之心配置.特效持续时间);
  UnitDamageTarget(上下文.施法单位, 上下文.目标单位, 咆哮之心配置.每跳伤害, false, false, ATTACK_TYPE_NORMAL, DAMAGE_TYPE_MIND, WEAPON_TYPE_WHOKNOWS);
}

export function 处理咆哮之心使用(this: void, 上下文: 物品技能事件上下文): void {
  debugLogForce("25．咆哮之心", "进入", "处理咆哮之心使用");

  if (!是否为咆哮之心(上下文.物品)) return;
  const 施法单位 = 上下文.施法单位;
  const 目标单位 = 上下文.目标单位;
  if (施法单位 == null || 施法单位 === 0 || 目标单位 == null || 目标单位 === 0) return;

  const 附加攻击 = R2I(GetUnitState(目标单位, ConvertUnitState(0x15))) / 咆哮之心配置.力量转攻击除数;
  SGSS_SetState(目标单位, 1, 附加攻击);
  const timer = CreateTimer();
  if (timer == null || timer === 0) return;
  咆哮之心表[GetHandleId(timer)] = { 施法单位, 目标单位, 附加攻击, 次数: 0 };
  TimerStart(timer, 咆哮之心配置.周期, true, on咆哮之心周期);
}

export {};

File: src/系统/02．物品系统/15．装备技能/00．物品/25．咆哮之心.lua
local ____lualib = require("lualib_bundle")
local __TS__Delete = ____lualib.__TS__Delete
local ____exports = {}
local ____01_FF0E_4E3B_52A8_6280_80FD_7269_54C1ID = require("系统.02．物品系统.15．装备技能.03．主动技能.00．公共.01．主动技能物品ID")
local _____5486_54EE_4E4B_5FC3_7269_54C1ID = ____01_FF0E_4E3B_52A8_6280_80FD_7269_54C1ID["咆哮之心物品ID"]
local ____00_FF0E_7269_54C1_4F7F_7528_89E6_53D1_914D_7F6E = require("系统.02．物品系统.15．装备技能.03．主动技能.03．物品使用触发.00．物品使用触发配置")
local _____5486_54EE_4E4B_5FC3_914D_7F6E = ____00_FF0E_7269_54C1_4F7F_7528_89E6_53D1_914D_7F6E["咆哮之心配置"]
local ____require_result_0 = require("lib.扩展函数.自定义扩展函数.03．调试输出")
local debugLogForce = ____require_result_0.debugLogForce
local jass = require("jass.common")
local ____require_result_1 = require("lib.扩展函数.封装函数.01．通用工具.03．特效")
local createTimedEffect = ____require_result_1.createTimedEffect
local ____require_result_2 = require("lib.扩展函数.Star扩展函数.00．SGSS")
local SGSS_SetState = ____require_result_2.SGSS_SetState
local GetItemTypeId = jass.GetItemTypeId
local GetUnitX = jass.GetUnitX
local GetUnitY = jass.GetUnitY
local GetUnitState = jass.GetUnitState
local ConvertUnitState = jass.ConvertUnitState
local R2I = jass.R2I
local UnitDamageTarget = jass.UnitDamageTarget
local CreateTimer = jass.CreateTimer
local TimerStart = jass.TimerStart
local GetExpiredTimer = jass.GetExpiredTimer
local GetHandleId = jass.GetHandleId
local DestroyTimer = jass.DestroyTimer
local ATTACK_TYPE_NORMAL = jass.ATTACK_TYPE_NORMAL
local DAMAGE_TYPE_MIND = jass.DAMAGE_TYPE_MIND
local WEAPON_TYPE_WHOKNOWS = jass.WEAPON_TYPE_WHOKNOWS
local _____5486_54EE_4E4B_5FC3_8868 = {}
local function _____662F_5426_4E3A_5486_54EE_4E4B_5FC3(_____7269_54C1)
    if _____7269_54C1 == nil or _____7269_54C1 == 0 then
        return false
    end
    return GetItemTypeId(_____7269_54C1) == _____5486_54EE_4E4B_5FC3_7269_54C1ID
end
local function ____on_5486_54EE_4E4B_5FC3_5468_671F()
    local timer = GetExpiredTimer()
    local timerID = GetHandleId(timer)
    local _____4E0A_4E0B_6587 = _____5486_54EE_4E4B_5FC3_8868[timerID]
    if _____4E0A_4E0B_6587 == nil then
        DestroyTimer(timer)
        return
    end
    if _____4E0A_4E0B_6587["次数"] >= _____5486_54EE_4E4B_5FC3_914D_7F6E["次数"] then
        SGSS_SetState(_____4E0A_4E0B_6587["目标单位"], 1, -_____4E0A_4E0B_6587["附加攻击"])
        __TS__Delete(_____5486_54EE_4E4B_5FC3_8868, timerID)
        DestroyTimer(timer)
        return
    end
    _____4E0A_4E0B_6587["次数"] = _____4E0A_4E0B_6587["次数"] + 1
    createTimedEffect(
        _____5486_54EE_4E4B_5FC3_914D_7F6E["特效路径"],
        GetUnitX(_____4E0A_4E0B_6587["目标单位"]),
        GetUnitY(_____4E0A_4E0B_6587["目标单位"]),
        0,
        _____5486_54EE_4E4B_5FC3_914D_7F6E["特效持续时间"]
    )
    UnitDamageTarget(
        _____4E0A_4E0B_6587["施法单位"],
        _____4E0A_4E0B_6587["目标单位"],
        _____5486_54EE_4E4B_5FC3_914D_7F6E["每跳伤害"],
        false,
        false,
        ATTACK_TYPE_NORMAL,
        DAMAGE_TYPE_MIND,
        WEAPON_TYPE_WHOKNOWS
    )
end
____exports["处理咆哮之心使用"] = function(_____4E0A_4E0B_6587)
    debugLogForce("25．咆哮之心", "进入", "处理咆哮之心使用")
    if not _____662F_5426_4E3A_5486_54EE_4E4B_5FC3(_____4E0A_4E0B_6587["物品"]) then
        return
    end
    local _____65BD_6CD5_5355_4F4D = _____4E0A_4E0B_6587["施法单位"]
    local _____76EE_6807_5355_4F4D = _____4E0A_4E0B_6587["目标单位"]
    if _____65BD_6CD5_5355_4F4D == nil or _____65BD_6CD5_5355_4F4D == 0 or _____76EE_6807_5355_4F4D == nil or _____76EE_6807_5355_4F4D == 0 then
        return
    end
    local _____9644_52A0_653B_51FB = R2I(GetUnitState(
        _____76EE_6807_5355_4F4D,
        ConvertUnitState(21)
    )) / _____5486_54EE_4E4B_5FC3_914D_7F6E["力量转攻击除数"]
    SGSS_SetState(_____76EE_6807_5355_4F4D, 1, _____9644_52A0_653B_51FB)
    local timer = CreateTimer()
    if timer == nil or timer == 0 then
        return
    end
    _____5486_54EE_4E4B_5FC3_8868[GetHandleId(timer)] = {["施法单位"] = _____65BD_6CD5_5355_4F4D, ["目标单位"] = _____76EE_6807_5355_4F4D, ["附加攻击"] = _____9644_52A0_653B_51FB, ["次数"] = 0}
    TimerStart(timer, _____5486_54EE_4E4B_5FC3_914D_7F6E["周期"], true, ____on_5486_54EE_4E4B_5FC3_5468_671F)
end
return ____exports
